src/dwarf/function.rs

```diff
@@ -112,6 +114,130 @@ where
+    pub(crate) name: Option<R<'dwarf>>,
+    pub(crate) call_file: Option<u64>,
+    pub(crate) call_line: u32,
+    pub(crate) call_column: u32,
+}
+
+
+struct InlinedFunctionAddress {
+    range: gimli::Range,
+    call_depth: usize,
+    /// An index into `Function::inlined_functions`.
+    function: usize,
+}
+
+
+pub(super) struct InlinedFunctions<'dwarf> {
+    /// List of all `DW_TAG_inlined_subroutine` details in this
+    /// function.
+    inlined_functions: Box<[InlinedFunction<'dwarf>]>,
+    /// List of `DW_TAG_inlined_subroutine` address ranges in this
+    /// function.
+    inlined_addresses: Box<[InlinedFunctionAddress]>,
+}
+
+impl<'dwarf> InlinedFunctions<'dwarf> {
+    pub(crate) fn parse(
+        dw_die_offset: gimli::UnitOffset<<R<'dwarf> as gimli::Reader>::Offset>,
+        unit: &gimli::Unit<R<'dwarf>>,
+        sections: &gimli::Dwarf<R<'dwarf>>,
+    ) -> Result<Self, Error> {
+        let mut entries = unit.entries_raw(Some(dw_die_offset))?;
+        let depth = entries.next_depth();
+        // The DIE offset we get is the one of the function. But we are
+        // interested in its children. So skip the necessary attributes.
+        let abbrev = entries.read_abbreviation().unwrap().unwrap();
+        debug_assert_eq!(abbrev.tag(), gimli::DW_TAG_subprogram);
+        let () = entries.skip_attributes(abbrev.attributes())?;
+
+        let mut inlined_functions = Vec::new();
+        let mut inlined_addresses = Vec::new();
+        Function::parse_children(
+            &mut entries,
+            depth,
+            unit,
+            sections,
+            &mut inlined_functions,
+            &mut inlined_addresses,
+            0,
+        )?;
+
+        // Sort ranges in "breadth-first traversal order", i.e. first by
+        // `call_depth` and then by `range.begin`. This allows finding
+        // the range containing an address at a certain depth using
+        // binary search.
+        // Note: Using DFS order, i.e. ordering by `range.begin` first
+        // and then by `call_depth`, would not work!
+        // Consider the two examples "[0..10 at depth 0], [0..2 at depth 1],
+        // [6..8 at depth 1]" and "[0..5 at depth 0], [0..2 at depth
+        // 1], [5..10 at depth 0], [6..8 at depth 1]".
+        // In this example, if you want to look up address 7 at depth 0,
+        // and you encounter [0..2 at depth 1], are you before or after
+        // the target range? You don't know.
+        inlined_addresses.sort_by(|r1, r2| {
+            if r1.call_depth < r2.call_depth {
+                Ordering::Less
+            } else if r1.call_depth > r2.call_depth {
+                Ordering::Greater
+            } else if r1.range.begin < r2.range.begin {
+                Ordering::Less
+            } else if r1.range.begin > r2.range.begin {
+                Ordering::Greater
+            } else {
+                Ordering::Equal
+            }
```

Comment:
isn't there some way to use tuple here? like `(r1.call_depth, r1.range.begin).cmp((r2.call_depth, r2.range.begin))` ?

Reply:
Yes, the comparison can be shortened.